pagespeed/envoy/envoy_fetch.cc

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "envoy_fetch.h"
+
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <list>
+#include <map>
+#include <set>
+
+#include "net/instaweb/http/public/async_fetch.h"
+#include "net/instaweb/http/public/inflating_fetch.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/condvar.h"
+#include "pagespeed/kernel/base/message_handler.h"
+#include "pagespeed/kernel/base/pool.h"
+#include "pagespeed/kernel/base/pool_element.h"
+#include "pagespeed/kernel/base/statistics.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/kernel/base/thread_system.h"
+#include "pagespeed/kernel/base/timer.h"
+#include "pagespeed/kernel/base/writer.h"
+#include "pagespeed/kernel/http/request_headers.h"
+#include "pagespeed/kernel/http/response_headers.h"
+#include "pagespeed/kernel/http/response_headers_parser.h"
+#include "pagespeed_remote_data_fetcher.h"
+#include "pagespeed/envoy/header_utils.h"
+
+namespace net_instaweb {
+
+// Default keepalive 60s.
+const int64 keepalive_timeout_ms = 60000;
+
+PagespeedDataFetcherCallback::PagespeedDataFetcherCallback(EnvoyFetch* fetch) { 
+  fetch_ = fetch;
+}
+
+void PagespeedDataFetcherCallback::onSuccess(Envoy::Http::MessagePtr& response) {
+  fetch_->setResponse(response->headers(), response->body());
+}
+
+void PagespeedDataFetcherCallback::onFailure(FailureReason reason) {
+  std::cout << "PagespeedDataFetcherCallback::onFailure\n";
+  std::cout.flush();
+}
+
+EnvoyFetch::EnvoyFetch(const GoogleString& url,
+                   AsyncFetch* async_fetch,
+                   MessageHandler* message_handler,
+                   EnvoyClusterManager& cluster_manager)
+    : str_url_(url),
+      fetcher_(NULL),
+      async_fetch_(async_fetch),
+      message_handler_(message_handler),
+      cluster_manager_(cluster_manager),
+      done_(false),
+      content_length_(-1),
+      content_length_known_(false) {
+}
+
+EnvoyFetch::~EnvoyFetch() {
+}
+
+void EnvoyFetch::FetchWithEnvoy() {
+  envoy::api::v2::core::HttpUri http_uri;
+  http_uri.set_uri("http://localhost:80");
+  http_uri.set_cluster("cluster1");
+  std::string uriHash("123456789");
```

Comment:
What's this?

Reply:
This variable name is not correct. Renaming it to **response_content_hash**. This hash value is response content hash (ignored by our PagespeedRemoteDataFetcher class) which just a dummy placeholder needed by envoy. 
Actually we referred RemoteDataFetcher code of envoy and [replicated here](https://github.com/apache/incubator-pagespeed-mod/blob/05e80e98faba31ff5a94f1244b1833fa88a1fae4/pagespeed/envoy/pagespeed_remote_data_fetcher.cc#L24). 
We kept the structure same as original. So passing dummy value.

Code updated here https://github.com/apache/incubator-pagespeed-mod/pull/1969/commits/ae9fdd05d8af8d99c3c9f7c31f2e3ef110f8eb04

Reply:
Is `const std::string& content_hash` really needed in the constructor of `PagespeedRemoteDataFetcher`? Can't we just remove it?

Reply:
We can remove this. Even earlier we planned to remove it. But I thought it to keep it just for the sake of envoy **RemoteDataFetcher** structure. 
As we are taking RemoteDataFetcher code as is from envoy and updating it just to remove response content hash check. So in future if RemoteDataFetcher structure gets updated, we can easily incorporate that in our code.
Let me know your views.

Reply:
If `PagespeedRemoteDataFetcher()` can hide the fact that it needs a content hash internally from consumers that seems nice to me; is there any downside to that?
(I'm not sure I understand why it needs to be exposed in the constructor, as that forces this detail upon consumers of `PagespeedRemoteDataFetcher`?).

Reply:
We can hide content hash part from consumers. I will make the change and push the code.
There is no downside to it except this is changing constructor signature of original class.
```
RemoteDataFetcher(Upstream::ClusterManager& cm, const ::envoy::api::v2::core::HttpUri& uri,
                    const std::string& content_hash, RemoteDataFetcherCallback& callback);
```

Reply:
Code pushed in https://github.com/apache/incubator-pagespeed-mod/pull/1969/commits/26fd5ca9c767bf7fa32a6561c40ceb79572dc6fb